docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,450 @@
+   - If a matching in-cluster object is selected, then it is injected as
+     follows:
+     - The `spec` field from the in-cluster resource is copied to the `spec`
+       field of the in-package resource (the injection point), overwriting it.
+     - An annotation with name `kpt.dev/injected-resource-name` and value set to
+       the name of the in-cluster resource is added (or overwritten) in the
+       in-package resource.
+1. The PackageVariant resource must have a condition of type `ConfigInjected`.
+   This will be set to `True` if and only if:
+   - All configuration injection processing is complete.
+   - There is no resource annotated as an injection point but having an invalid
+     annotation value (i.e., other than `required` or `optional`).
+   - Matching in-cluster resources were successfully injected for all `required`
+     injection points (all required injection point ConditionTypes are `True`).
+   - There are no ambiguous condition types due to conflicting GVK and name
+     values. These must be disambiguated in the upstream package, if so.
+
+Note that by allowing the use of GVK, not just name, in the selector, more
+precision in selection is enabled. This is a way to constrain the injections
+that will be done. That is, if the package has 10 different objects with
+`config-injection` annotation, the PackageVariant could say it only wants to
+replace certain GVKs, allowing better control.
+
+Consider, for example, if the cluster contains these resources:
+
+- GVK1 foo
+- GVK1 bar
+- GVK2 foo
+- GVK2 bar
+
+If we could only define injection selectors based upon name, it would be
+impossible to ever inject one GVK with `foo` and another with `bar`. Instead,
+by using GVK, we can accomplish this with a list of selectors like:
+
+ - GVK1 foo
+ - GVK2 bar
+
+That said, often name will be sufficiently unique when combined with the
+in-package resource GVK, and so making the selector GVK optional is more
+convenient.
+
+#### Order of Mutation During Creation
+Note that the KRM function pipeline defined in the Kptfile runs every time Porch
+saves the package. Thus, the Draft package will undergo mutations in this order
+during creation:
+
+1. Package Context Injections
+1. Kptfile KRM Function Pipeline
+1. PackageVariant KRM Function Pipeline
+1. Kptfile KRM Function Pipeline
+1. Config Injection
+1. Kptfile KRM Function Pipeline
+
+#### Order of Mutation During Updates
+
+Updates to the downstream PackageRevision can happen for several different
+reasons:
+
+1. An injected in-cluster object is updated.
+1. The PackageVariant resource is updated, which could change any of the options
+   for introducing variance, or could also change the upstream package revision
+   referenced.
+1. A new revision of the upstream package has been selected due to a floating
+   tag change, or due to a force retagging of the upstream.
+
+TODO(johnbelamaric): Figure out and document how the mutation process works in
```

Comment:
@natasha41575 any thoughts on this?

Reply:
I agree with the description from @natasha41575 of what should happen here. I think in all three situations, we want to update the package. In practice this means updating a draft if one exists, or create a new draft from the latest published revision if no draft exists.

Reply:
Ok, changing the way the KRM function pipeline works, to just modify the Kptfile pipeline, made this way simpler. Updated.